### PR TITLE
Add general addObject() to FEProblemBase

### LINKED ISSUE #15484: Add method to FEProblemBase to add objects to TheWarehouse
## Reason
As we transition to using `TheWarehouse` for object storage, there should be a general templated method for adding objects to said warehouse.

## Design
Add `FEProblemBase::addObject()`, which will call `Factory::create()` and add objects to the warehouse.

## Impact
Added API. Will change internal APIs that add objects to `FEProblemBase::theWarehouse()` to use it.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #15484

### CODE REVIEW

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -2695,62 +2693,17 @@ FEProblemBase::addFVKernel(const std::string & fv_kernel_name,
...
-    parameters.set<SubProblem *>("_subproblem") = this;
-    parameters.set<SystemBase *>("_sys") = _nl.get();
-  }
-
-  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
-  {
-    std::shared_ptr<FVKernel> k = _factory.create<FVKernel>(fv_kernel_name, name, parameters, tid);
-    theWarehouse().add(k);
-  }
+  // TODO: this changes an overridden use_displaced_mesh - do we really need it for FV?
```
Comment (aeslaughter) [inline framework/src/problems/FEProblemBase.C:2696]: @rwcarlsen @lindsayad What do you say to this?

Comment (loganharbour) [reply framework/src/problems/FEProblemBase.C:2696]: This was in there before - I just lazily copy-pasta. 

We do it for standard Kernels too, so I'd say it should probably be there, too?

**Comment by loganharbour** [comment]
There's also cases where we sometimes set `_sys` as well.

I'm thinking that it may be appropriate here to set it, too.

**Comment by aeslaughter** [comment]
> There's also cases where we sometimes set `_sys` as well.
> 
> I'm thinking that it may be appropriate here to set it, too.

Is it worth having flags for adding the various things. Or an Optional MultiMooseEnum?

**Comment by loganharbour** [comment]
> Or an Optional MultiMooseEnum?

Can you explain this?

**Comment by aeslaughter** [comment]
> Can you explain this?

Maybe a MultiMooseEnum isn't right, but the idea I am trying to capture is something like this.

```c++
addObject(type, name, params, "sys", "fe_problem");
```

Internally, the addObject could check the supplied items and add the ones requested. Of course we could have FEProblem or whatever as default.

Probably overkill.

**Comment by loganharbour** [comment]
> Probably overkill.

I think so. If we manage the const-ness in the objects themselves, it's okay to provide a little too much information.

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (rwcarlsen) [reply framework/src/problems/FEProblemBase.C:2693]: I think this looks fine.

Comment (loganharbour) [reply framework/src/problems/FEProblemBase.C:2693]: So - good to remove the TODO comment, then?

Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:2693]: Nice. Yea!

File: framework/include/problems/FEProblemBase.h [hunk 2]
```diff
@@ -766,6 +766,37 @@ class FEProblemBase : public SubProblem, public Restartable
...
+    std::vector<std::shared_ptr<T>> objects(n_threads);
+    for (THREAD_ID tid = 0; tid < n_threads; ++tid)
+    {
+      std::shared_ptr<T> obj = _factory.create<T>(type, name, parameters, tid);
+      theWarehouse().add(obj);
+      objects[tid] = std::move(obj);
+    }
+
+    return objects;
+  }
```
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:798]: I would prefer that this body be outside the class definition. Takes up too much space for a developer perusing through APIs

**Review by lindsayad — DISMISSED** [review]

File: framework/include/problems/FEProblemBase.h (hunk 2 above)
Comment (loganharbour) [reply framework/include/problems/FEProblemBase.h:798]: I'm not sure how I would go about doing that with this being a templated method.

Comment (loganharbour) [reply framework/include/problems/FEProblemBase.h:798]: Oh - disregard. I follow. Sounds good.

**Review by aeslaughter — DISMISSED** [review]

**Comment by lindsayad** [comment]
Seems like the Grizzly failure is real/related to this PR?

**Comment by loganharbour** [comment]
> Seems like the Grizzly failure is real/related to this PR?

Damn - I think so. Let's see...

**Comment by bwspenc** [comment]
@loganharbour Yes, I do think that failure is caused by this PR. I call FEProblem::addDistribution() from a custom sampler in Grizzly. It's not immediately obvious to me why it doesn't work the same as it did before. If we really need to, I can rewrite that code so that I don't do it that way.

**Comment by loganharbour** [comment]
@bwspenc that shouldn't be necessary - this should be working at the framework level. I'm glad an application caught the issue!

**Comment by loganharbour** [comment]
Grizzly is good:
```
Ran 375 tests in 16.9 seconds.
375 passed, 5 skipped, 0 pending, 0 failed
```

Silly mistake.

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**